### PR TITLE
aux/fixed connor

### PR DESCRIPTION
**Pull request by DoctorCotic**
Fixed version of the bot.

### CODE REVIEW

File: connor/chargeOrders.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+			Benchmarks: benchmarks,
+			Network: &sonm.BidNetwork{
+				Overlay:  false,
+				Outbound: true,
+				Incoming: false,
+			},
+		},
+	})
+	if err != nil {
+		t.c.logger.Warn("cannot create bidOrder:", zap.String(err.Error(), "w"))
```
Comment (sshaman1101) [inline connor/chargeOrders.go]: [Wat?](https://www.youtube.com/watch?v=MCn9lL94sxQ)

File: connor/advancedPoolModule.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+				} // try to get average data from nanopool.org
+
+				if w.WorkerAvgHashrate == 0.0 {
+					p.c.logger.Info("worker reported and average hashrate = 0 send to Connor's blacklist", zap.Int64("worker id :", w.DealID))
+					if err := p.SendToConnorBlackList(ctx, dealInfo); err != nil {
+						return err
+					}
+					continue
+				} else {
+					iterationForAccessAvg := int64(4)
```
Comment (sshaman1101) [inline connor/advancedPoolModule.go]:
```
>>> len('iterationForAccessAvg')
21
>>> len('int64(4)')
8
```

so why to introduce a new variable and then use it only once?

File: cmd/connor/main.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func main() {
+	cmd.NewCmd("connor", appVersion, &configFlag, &versionFlag, run).Execute()
+}
+
+func run() error {
+	cfg, err := connor.NewConfig(configFlag)
+	if err != nil {
+		return fmt.Errorf("cannot load config file: %s\r\n", err)
```
Comment (3Hren) [inline cmd/connor/main.go]: No \r\n in errors please.

File: cmd/connor/main.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	key, err := cfg.Eth.LoadKey()
+	if err != nil {
+		return fmt.Errorf("cannot open keys: %v\r\n", err)
```
Comment (3Hren) [inline cmd/connor/main.go]: No \r\n in errors please.

File: cmd/connor/main.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	logger := logging.BuildLogger(cfg.Log.LogLevel())
+	ctx := log.WithLogger(context.Background(), logger)
+
+	c, err := connor.NewConnor(ctx, key, cfg)
```
Comment (3Hren) [inline cmd/connor/main.go]: Why is it required to pass ctx here?

File: connor/advancedPoolModule.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+//Tracking hashrate with using Connor's blacklist. Get data for 1 hour and another time => Detecting deviation.
+func (p *PoolModule) AdvancedPoolHashrateTracking(ctx context.Context, reportedPool watchers.PoolWatcher, avgPool watchers.PoolWatcher) error {
+	workers, err := p.c.db.GetWorkersFromDB()
+	if err != nil {
+		p.c.logger.Error("cannot get worker from pool DB", zap.Error(err))
+		return err
+	}
+
+	for _, w := range workers {
+		if w.BadGuy > 5 {
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: So magic.

File: connor/advancedPoolModule.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+		if w.Iterations == 0 {
+			newIteration := w.Iterations + 1
+
+			if err := p.c.db.UpdateIterationPoolDB(newIteration, w.DealID); err != nil {
+				return err
+			}
+			continue
+		}
+
+		p.c.logger.Info("Update iteration worker less than 4", zap.Int64("worker", w.DealID))
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: Lowerscore.

File: connor/advancedPoolModule.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+		if err != nil {
+			return err
+		}
+
+		if w.Iterations < numberOfIterationsForH1 {
+			if err := p.UpdatePoolData(ctx, reportedPool, p.c.cfg.PoolAddress.EthPoolAddr, int64(PoolTypeReportedHashrate)); err != nil {
+				return err
+			}
+
+			if w.WorkerReportedHashrate == 0.0 {
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: Comparing floats with zero is [tricky](https://bitbashing.io/comparing-floats.html).

File: connor/advancedPoolModule.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+				if err := p.UpdatePoolData(ctx, avgPool, p.c.cfg.PoolAddress.EthPoolAddr+"/1", int64(PoolTypeAvgHashrate)); err != nil {
+					return err
+				} // try to get average data from nanopool.org
+
+				if w.WorkerAvgHashrate == 0.0 {
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: Here too.

File: connor/advancedPoolModule.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+				if err := p.SendToConnorBlackList(ctx, dealInfo); err != nil {
+					return err
+				}
+				continue
+			}
+
+			workerAvgHashrate := uint64(w.WorkerAvgHashrate * hashes)
+			p.c.logger.Info("update avg hashrate", zap.Uint64("AVG", workerAvgHashrate), zap.Int64("worker", w.DealID))
+
+			changeResponseAvgWorker := float64(100 - (float64(workerAvgHashrate*100) / float64(bidHashrate)))
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: Code dup.

File: connor/advancedPoolModule.go [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+		}
+
+	}
+	return nil
+}
+
+//Detects the percentage of deviation of the hashrate and save SupplierID (by MasterID) to Connor's blacklist .
+func (p *PoolModule) AdvancedDetectingDeviation(ctx context.Context, changePercentDeviationWorker float64, worker *database.PoolDb, dealInfo *sonm.DealInfoReply) error {
+
+	if changePercentDeviationWorker >= 20 {
```
Comment (3Hren) [inline connor/advancedPoolModule.go]: Magic.

File: connor/defaultPoolModule.go [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+			newStatus := worker.BadGuy + 1
+			err := p.c.db.UpdateBadGayStatusInPoolDB(worker.DealID, newStatus, time.Now());
+			if err != nil {
+				return err
+			}
+		} else {
+			if err := p.DestroyDeal(ctx, dealInfo); err != nil {
+				return err
+			}
+			err := p.c.db.UpdateBadGayStatusInPoolDB(worker.DealID, int64(BanStatusWORKERINPOOL), time.Now());
```
Comment (sshaman1101) [inline connor/defaultPoolModule.go]: Hmmm, I exactly can't find »WORKERINPOOL« in the Oxford dictionary.

File: connor/defaultPoolModule.go [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+	PoolTypeReportedHashrate PoolType = 0
+	PoolTypeAvgHashrate      PoolType = 1
+)
+
+type BanStatus int32
+
+const (
+	BanStatusBANNED       BanStatus = 1
+	BanStatusMASTERBAN    BanStatus = 2
+	BanStatusWORKERINPOOL BanStatus = 6
```
[the remark this answers is not in the document]
Comment (DoctorCotic) [reply connor/defaultPoolModule.go]: >_<" I'm dumb

File: connor/defaultPoolModule.go [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+	reply, err := p.c.TaskClient.Start(ctx, startTaskRequest)
+	if err != nil {
+		p.c.logger.Info("cannot create start task with given container - deal closed", zap.Any("deal", deal))
+
+		err = p.c.db.UpdateDeployStatusDealInDB(deal.Id.Unwrap().Int64(), int64(DeployStatusDESTROYED))
+		if err != nil {
+			return nil, err
+		}
+		if err = p.c.db.UpdateDealStatusDb(deal.Id.Unwrap().Int64(), sonm.DealStatus_DEAL_CLOSED); err !=nil{
```
[the remark this answers is not in the document]
Comment (DoctorCotic) [reply connor/defaultPoolModule.go]: You are abs right. I'm dumb (2)

File: connor/defaultPoolModule.go [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+	return reply, nil
+}
+
+// Checks for a deal in the worker list. If it is not there, adds.
+func (p *PoolModule) AddWorkerToPoolDB(ctx context.Context, deal *sonm.DealInfoReply, addr string) error {
+	val, err := p.c.db.GetWorkerFromPoolDb(deal.Deal.Id.String())
+	if err != nil {
+		return err
+	}
+	if val == deal.Deal.Id.String() {
```
[the remark this answers is not in the document]
Comment (sshaman1101) [reply connor/defaultPoolModule.go]: No, this checks that function returns an ID that you're awaiting for, but the method also can return the magic string `already in Pool!` (see my comment above). Refactor that method and remove this (almost meaningless) check, please.

File: connor/defaultPoolModule.go [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+		if iteration < numberOfIterationsForH1 {
+			if err = p.UpdatePoolData(ctx, reportedPool, p.c.cfg.PoolAddress.EthPoolAddr, int64(PoolTypeReportedHashrate)); err != nil {
+				return err
+			}
+			changePercentRHWorker := 100 - ((uint64(w.WorkerReportedHashrate*hashes) * 100) / bidHashrate)
+			if err = p.DetectingDeviation(ctx, changePercentRHWorker, w, dealInfo); err != nil {
+				return err
+			}
+		} else {
+			err := p.UpdatePoolData(ctx, avgPool, p.c.cfg.PoolAddress.EthPoolAddr+"/1", int64(PoolTypeAvgHashrate))
```
[the remark this answers is not in the document]
Comment (DoctorCotic) [reply connor/defaultPoolModule.go]: Really need a constant here? This value is needed to correctly access the nanopool API.

File: connor/database/db.go [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+	defer rows.Close()
+	for rows.Next() {
+		var dealID string
+		err = rows.Scan(&dealID)
+		if err != nil {
+			return "", err
+		}
+		return dealID, nil
+	}
+	return "already in Pool!", nil
```
[the remark this answers is not in the document]
Comment (sshaman1101) [reply connor/database/db.go]: nil, error.New("deal not found")

File: connor/checkTask.go [hunk 18]
```diff
@@ -69,11 +69,7 @@ func (p *PoolModule) CheckTaskStatus(ctx context.Context) error {
 				}
 
 			case sonm.DealStatus_DEAL_CLOSED:
-				err = p.c.db.UpdateDeployStatusDealInDB(d.DealID, int64(DeployStatusDestroyed))
-				if err != nil {
-					return err
-				}
-				if err = p.c.db.UpdateDealStatusDb(d.DealID, sonm.DealStatus_DEAL_CLOSED); err != nil {
+				if err = p.c.db.UpdateDeployAndDealStatusDb(d.DealID, int64(DeployStatusDestroyed), sonm.DealStatus_DEAL_CLOSED); err != nil {
```
Comment (sshaman1101) [inline connor/checkTask.go]: Db -> DB

File: connor/defaultPoolModule.go [hunk 19]
```diff
@@ -80,12 +80,12 @@ func (p *PoolModule) DeployNewContainer(ctx context.Context, cfg *Config, deal *
...
 			if err != nil {
-				return nil, fmt.Errorf("fail finish deal %v", err)
+				return nil, fmt.Errorf("fail finish deal: %s", err)
 			}
-			p.c.logger.Info("fail start task :: deal finished", zap.Int64("deal", deal.Id.Unwrap().Int64()))
+			p.c.logger.Info("fail start task deal finished", zap.Int64("deal", deal.Id.Unwrap().Int64()))
 
 		case sonm.DealStatus_DEAL_CLOSED:
-			p.c.logger.Info("fail finish deal :: already finished from worker", zap.Any("deal", deal))
+			p.c.logger.Info("deal already finished from worker", zap.Any("deal", deal))
```
Comment (sshaman1101) [inline connor/defaultPoolModule.go]: my master

File: connor/advancedPoolModule.go [hunk 20]
```diff
@@ -101,29 +99,29 @@ func (p *PoolModule) AdvancedPoolHashrateTracking(ctx context.Context, reportedP
...
 
 			changeResponseAvgWorker := float64(100 - (float64(workerAvgHashrate*100) / float64(bidHashrate)))
 			if changeResponseAvgWorker <= 0 {
-				p.c.logger.Info("change response reported hashrate worker < 0. It's ok.", zap.Int64("worker", w.DealID))
+				p.c.logger.Info("change response reported hashrate worker > 0. It's ok.", zap.Int64("worker", w.DealID))
 				continue
 			}
 
-			p.c.logger.Info("Pool inf :: worker deviation (average data)", zap.Int64("iteration", w.Iterations),
+			p.c.logger.Info("pool inf :: worker deviation (average data)", zap.Int64("iteration", w.Iterations),
```
Comment (sshaman1101) [inline connor/advancedPoolModule.go]: inf is for infinity?

File: connor/traderModule.go [hunk 21]
```diff
@@ -184,7 +181,7 @@ func (t *TraderModule) deployedDealProfitTracking(ctx context.Context, actualPri
 
 	changePricePercent, _ := big.NewFloat(0).SetInt64(div.Int64()).Float64()
 
-	if changePricePercent > 100+t.c.cfg.Sensitivity.DealsChangePercent {
+	if changePricePercent > 100+t.c.cfg.Trade.DealsChangePercent {
```
Comment (sshaman1101) [inline connor/traderModule.go]: yet another magic number?

File: connor/traderModule.go [hunk 22]
```diff
@@ -441,16 +437,21 @@ func (t *TraderModule) CancelAllOrders(ctx context.Context) error {
...
-			t.c.db.UpdateOrderInDB(o.OrderID, int64(OrderStatusCANCELLED))
+			t.c.db.UpdateOrderInDB(o.OrderID, int64(OrderStatusCancelled))
 		}
 	}
 	return nil
 }
+
+// change the price depending on the set hashrate from ask
+func (t *TraderModule) ChangePriceOfAskHashrate(ctx context.Context){
+
```
Comment (sshaman1101) [inline connor/traderModule.go]: wow, invisible code!

File: etc/connor.yaml [hunk 23]
```diff
@@ -0,0 +1,114 @@
+log:
+  # The desired logging level.
+  # Allowed values are "debug", "info", "warn", "error", "panic" and "fatal"
+  level: debug
+
+# Node address.
+market:
+   endpoint: "any@127.0.0.1:15030"
```
Comment (sshaman1101) [inline etc/connor.yaml]: any is not the best choice for ethereum address placeholder

File: connor/emergency.go [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+			}
+			t.c.logger.Info("order id cancelled", zap.Int64("order", o.OrderID))
+			t.c.db.UpdateOrderInDB(o.OrderID, int64(OrderStatusCancelled))
+		}
+	}
+	return nil
+}
+
+func (t *EmergencyModule) CloseAllActiveDeals(ctx context.Context) error {
+	return nil
```
Comment (sshaman1101) [inline connor/emergency.go]: wow, so method, such empty

File: connor/advancedPoolModule.go (hunk 20 above)
Comment (DoctorCotic) [reply connor/advancedPoolModule.go]: inf is for INFERNO

**Pull request closed without merging**